### PR TITLE
refactor: improve input validation with proper exceptions

### PR DESCRIPTION
**Pull request by nh13**
# Validation Improvements and Comprehensive Testing

This PR improves error handling and adds comprehensive validation tests, building on top of the recent attrs-to-dataclasses migration and superintervals adoption in PR #54.

---

### 1. Replace assert with ValueError for User Input Validation
**Commit:** `refactor: replace assert with ValueError for user input validation` (3e3616d)

**Rationale:** Following Python best practices, `assert` statements should be used for internal invariants (conditions that should never fail in correct code), while `ValueError` should be used for invalid user input that can reasonably occur.

**Changes to pybedlite/bed_source.py:**
- `open()`: ValueError if file handle is closed when pre-opened filehandle specified  
- `close()`: ValueError if attempting to close an unopened file
- `parse_rgb()`: ValueError if RGB value doesn't have exactly 3 comma-separated integers
- `__iter__()`: ValueError if BED record has fewer than 3 required fields

**Changes to pybedlite/bed_writer.py:**
- `__init__()`: ValueError if num_fields not in valid range (3-12)
- `open()`: ValueError if file handle is closed when pre-opened filehandle specified
- `close()`: ValueError if attempting to close an unopened file

All ValueError exceptions include descriptive error messages with context to help users understand and fix validation failures. Docstrings were improved with complete `Args`, `Returns`, and `Raises` sections following Google style guide.

**Example Improvement:**
```python
# Before
assert (
    len(rgb_split) == 3
), "item_rgb, if defined, must contain 3 comma separated integers"

# After
if len(rgb_split) != 3:
    raise ValueError(
        f"item_rgb, if defined, must contain 3 comma separated integers, "
        f"got {len(rgb_split)}"
    )
```

**Files Changed:**
- `pybedlite/bed_source.py`
- `pybedlite/bed_writer.py`

---

### 2. Add Comprehensive Validation Tests
**Commit:** `test: add comprehensive validation tests to improve coverage` (6f00f5a)

Added 18 new validation tests covering error paths and edge cases in BedRecord, BedSource, and BedWriter classes to ensure the ValueError-based validation logic works correctly.

**BedRecord Validation Tests (12):**
- `test_bedrecord_validation_start_end` - validates start < end (both equal and greater-than cases)
- `test_bedrecord_validation_thick_start_end_mismatch` - validates thick_start/thick_end must both be defined or both None
- `test_bedrecord_validation_block_count_without_sizes` - validates block fields must be defined together
- `test_bedrecord_validation_block_count_without_starts` - validates block fields must be defined together
- `test_bedrecord_validation_block_sizes_without_count` - validates block fields must be defined together
- `test_bedrecord_validation_block_starts_without_count` - validates block fields must be defined together
- `test_bedrecord_validation_block_sizes_length_mismatch` - validates block_sizes length matches block_count
- `test_bedrecord_validation_block_starts_length_mismatch` - validates block_starts length matches block_count
- `test_bedrecord_validation_block_starts_first_nonzero` - validates first block_start must be 0
- `test_bedrecord_validation_block_end_mismatch` - validates last block end must equal interval end
- `test_bedrecord_as_bed_line_invalid_field_count` - validates num_fields must be in range 3-12
- `test_bedrecord_bed_field_num_coverage` - covers all branches of bed_field_num property for BED3, BED4, BED5, BED6, BED8, BED9, BED12

**BedSource Validation Tests (3):**
- `test_bedsource_close_unopened_file` - validates file state before closing
- `test_bedsource_invalid_rgb` - validates RGB parsing with wrong number of components
- `test_bedsource_too_few_fields` - validates BED records must have at least 3 fields

**BedWriter Validation Tests (3):**
- `test_bedwriter_invalid_num_fields_too_small` - validates num_fields >= 3
- `test_bedwriter_invalid_num_fields_too_large` - validates num_fields <= 12
- `test_bedwriter_close_unopened_file` - validates file state before closing

All tests use `pytest.raises()` to verify exceptions are raised with expected message patterns, preventing regressions in validation logic.

**Files Changed:**
- `pybedlite/tests/test_pybedlite.py` (new file with 18 tests)

---

## 📝 Commit Summary

All commits follow conventional commit format:

```
6f00f5a test: add comprehensive validation tests to improve coverage
3e3616d refactor: replace assert with ValueError for user input validation
```

---

## 📦 Files Changed Summary

- `pybedlite/bed_source.py` - ValueError migration + improved docstrings
- `pybedlite/bed_writer.py` - ValueError migration + improved docstrings  
- `pybedlite/tests/test_pybedlite.py` - 18 new validation tests (new file)

---

## ✅ Testing & Validation

- ✅ All tests should pass across Python 3.10, 3.11, 3.12, 3.13 (Python 3.9 dropped in PR #54)
- ✅ Black formatting applied
- ✅ All changes maintain backward compatibility
- ✅ No breaking changes to public APIs
- ✅ Improved error messages help users fix validation failures

---

## 📚 Documentation

All modified functions include improved docstrings with:
- Standardized Google-style format
- Complete `Args`, `Returns`, `Raises` sections
- Better parameter descriptions with types
- Clear explanations of validation rules

---

## 🔗 Related Work

This PR builds on:
- PR #54: Dropped Python 3.9 support and removed attrs dependency
- PR #42: Migrated from cgranges to superintervals

The validation improvements in this PR complement the recent refactorings by ensuring robust error handling and comprehensive test coverage for the modernized codebase.

### CODE REVIEW

File: pybedlite/bed_source.py [hunk 1]
```diff
@@ -127,21 +143,19 @@ def parse_rgb(rgb: str) -> Tuple[int, int, int]:
             self.open()
             self._file_is_open = True
 
-        assert self._file_is_open and self._in_fh is not None, (
-            "File must be opened before iterating over it!"
-        )
-
+        assert self._in_fh is not None, "Assertion present to satisfy mypy"
```
Comment (clintval) [inline pybedlite/bed_source.py]: The previous exception message was better.
